File: Sources/Domain/TrafficRateAggregator.swift
struct TrafficRateAggregator: Sendable {
  private let windowDuration: Double
  private let smoothingWindowCount: Int
  private var accumulatedDuration = 0.0
  private var accumulatedKernel = TrafficBytes.zero
  private var accumulatedCategories = CategorizedTrafficBytes.zero
  private var recentRawRates: [CategorizedTrafficRates] = []

  init(windowDuration: Double = 1, smoothingWindowCount: Int = 2) {
    precondition(windowDuration > 0)
    precondition(smoothingWindowCount > 0)
    self.windowDuration = windowDuration
    self.smoothingWindowCount = smoothingWindowCount
  }

  mutating func consume(
    _ report: TrafficDeltaReport,
    elapsedSeconds: Double
  ) -> TrafficRateWindow? {
    guard elapsedSeconds.isFinite, elapsedSeconds > 0 else {
      return nil
    }

    accumulatedDuration += elapsedSeconds
    accumulatedKernel = accumulatedKernel + report.kernel
    accumulatedCategories = add(accumulatedCategories, report.categories)

    guard accumulatedDuration >= windowDuration else {
      return nil
    }

    let aggregate = TrafficDeltaReport(
      kernel: accumulatedKernel,
      categories: accumulatedCategories
    )
    let raw = rates(from: accumulatedCategories, duration: accumulatedDuration)
    recentRawRates.append(raw)
    if recentRawRates.count > smoothingWindowCount {
      recentRawRates.removeFirst(recentRawRates.count - smoothingWindowCount)
    }

    let window = TrafficRateWindow(
      raw: raw,
      smoothed: average(recentRawRates),
      coverage: aggregate.coverage
    )
    accumulatedDuration = 0
    accumulatedKernel = .zero
    accumulatedCategories = .zero
    return window
  }

  mutating func reset() {
    accumulatedDuration = 0
    accumulatedKernel = .zero
    accumulatedCategories = .zero
    recentRawRates = []
  }

  private func rates(
    from categories: CategorizedTrafficBytes,
    duration: Double
  ) -> CategorizedTrafficRates {
    CategorizedTrafficRates(
      proxy: rate(from: categories.proxy, duration: duration),
      direct: rate(from: categories.direct, duration: duration),
      reject: rate(from: categories.reject, duration: duration),
      unknown: rate(from: categories.unknown, duration: duration)
    )
  }

  private func rate(from bytes: TrafficBytes, duration: Double) -> TrafficRate {
    TrafficRate(
      uploadBytesPerSecond: bytesPerSecond(bytes.upload, duration: duration),
      downloadBytesPerSecond: bytesPerSecond(bytes.download, duration: duration)
    )
  }

  private func bytesPerSecond(_ bytes: UInt64, duration: Double) -> UInt64 {
    let value = Double(bytes) / duration
    guard value < Double(UInt64.max) else {
      return UInt64.max
    }
    return UInt64(value)
  }

  private func average(_ rates: [CategorizedTrafficRates]) -> CategorizedTrafficRates {
    CategorizedTrafficRates(
      proxy: average(rates.map(\.proxy)),
      direct: average(rates.map(\.direct)),
      reject: average(rates.map(\.reject)),
      unknown: average(rates.map(\.unknown))
    )
  }

  private func average(_ rates: [TrafficRate]) -> TrafficRate {
    let count = UInt64(rates.count)
    guard count > 0 else {
      return .zero
    }

    let totals = rates.reduce(TrafficBytes.zero) {
      $0
        + TrafficBytes(
          upload: $1.uploadBytesPerSecond,
          download: $1.downloadBytesPerSecond
        )
    }
    return TrafficRate(
      uploadBytesPerSecond: totals.upload / count,
      downloadBytesPerSecond: totals.download / count
    )
  }

  private func add(
    _ left: CategorizedTrafficBytes,
    _ right: CategorizedTrafficBytes
  ) -> CategorizedTrafficBytes {
    CategorizedTrafficBytes(
      proxy: left.proxy + right.proxy,
      direct: left.direct + right.direct,
      reject: left.reject + right.reject,
      unknown: left.unknown + right.unknown
    )
  }
}

File: platform/windows/MihomoMeter.Windows.Core/Domain/TrafficRateAggregator.cs
namespace MihomoMeter.Windows.Core.Domain;

public sealed class TrafficRateAggregator
{
    private readonly double _windowDuration;
    private readonly int _smoothingWindowCount;
    private readonly Queue<CategorizedTrafficRates> _recentRates = new();
    private double _accumulatedDuration;
    private TrafficBytes _accumulatedKernel = TrafficBytes.Zero;
    private CategorizedTrafficBytes _accumulatedBytes = CategorizedTrafficBytes.Zero;

    public TrafficRateAggregator(double windowDuration = 1, int smoothingWindowCount = 2)
    {
        if (!double.IsFinite(windowDuration) || windowDuration <= 0)
        {
            throw new ArgumentOutOfRangeException(nameof(windowDuration));
        }

        if (smoothingWindowCount <= 0)
        {
            throw new ArgumentOutOfRangeException(nameof(smoothingWindowCount));
        }

        _windowDuration = windowDuration;
        _smoothingWindowCount = smoothingWindowCount;
    }

    public TrafficRateWindow? Consume(TrafficDeltaReport report, double elapsedSeconds)
    {
        if (!double.IsFinite(elapsedSeconds) || elapsedSeconds <= 0)
        {
            return null;
        }

        _accumulatedDuration += elapsedSeconds;
        _accumulatedKernel = TrafficBytes.Add(_accumulatedKernel, report.Kernel);
        _accumulatedBytes = new CategorizedTrafficBytes(
            TrafficBytes.Add(_accumulatedBytes.Proxy, report.Categories.Proxy),
            TrafficBytes.Add(_accumulatedBytes.Direct, report.Categories.Direct),
            TrafficBytes.Add(_accumulatedBytes.Reject, report.Categories.Reject),
            TrafficBytes.Add(_accumulatedBytes.Unknown, report.Categories.Unknown));

        if (_accumulatedDuration < _windowDuration)
        {
            return null;
        }

        var aggregate = new TrafficDeltaReport(_accumulatedKernel, _accumulatedBytes);
        var raw = RatesFrom(_accumulatedBytes, _accumulatedDuration);
        _recentRates.Enqueue(raw);
        while (_recentRates.Count > _smoothingWindowCount)
        {
            _recentRates.Dequeue();
        }

        var smoothed = Average(_recentRates);
        _accumulatedDuration = 0;
        _accumulatedKernel = TrafficBytes.Zero;
        _accumulatedBytes = CategorizedTrafficBytes.Zero;
        return new TrafficRateWindow(raw, smoothed, aggregate.Coverage);
    }

    public void Reset()
    {
        _accumulatedDuration = 0;
        _accumulatedKernel = TrafficBytes.Zero;
        _accumulatedBytes = CategorizedTrafficBytes.Zero;
        _recentRates.Clear();
    }

    private static CategorizedTrafficRates RatesFrom(
        CategorizedTrafficBytes bytes,
        double duration)
    {
        return new CategorizedTrafficRates(
            RateFrom(bytes.Proxy, duration),
            RateFrom(bytes.Direct, duration),
            RateFrom(bytes.Reject, duration),
            RateFrom(bytes.Unknown, duration));
    }

    private static TrafficRate RateFrom(TrafficBytes bytes, double duration)
    {
        return new TrafficRate(
            BytesPerSecond(bytes.Upload, duration),
            BytesPerSecond(bytes.Download, duration));
    }

    private static ulong BytesPerSecond(ulong bytes, double duration)
    {
        var value = bytes / duration;
        return value >= ulong.MaxValue ? ulong.MaxValue : (ulong)value;
    }

    private static CategorizedTrafficRates Average(IEnumerable<CategorizedTrafficRates> rates)
    {
        var items = rates.ToArray();
        return new CategorizedTrafficRates(
            Average(items.Select(item => item.Proxy)),
            Average(items.Select(item => item.Direct)),
            Average(items.Select(item => item.Reject)),
            Average(items.Select(item => item.Unknown)));
    }

    private static TrafficRate Average(IEnumerable<TrafficRate> rates)
    {
        var items = rates.ToArray();
        if (items.Length == 0)
        {
            return TrafficRate.Zero;
        }

        var total = items.Aggregate(
            TrafficBytes.Zero,
            (current, rate) => TrafficBytes.Add(
                current,
                new TrafficBytes(rate.UploadBytesPerSecond, rate.DownloadBytesPerSecond)));
        var count = (ulong)items.Length;
        return new TrafficRate(total.Upload / count, total.Download / count);
    }
}
